clang/lib/CConv/DeclRewriter.cpp

```diff
@@ -0,0 +1,410 @@
+//=--DeclRewriter.cpp---------------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+//===----------------------------------------------------------------------===//
+#include "clang/CConv/CCGlobalOptions.h"
+#include "clang/CConv/RewriteUtils.h"
+#include "clang/CConv/Utils.h"
+#include "llvm/Support/raw_ostream.h"
+#include "clang/CConv/DeclRewriter.h"
+#include "clang/AST/Decl.h"
+#include "clang/AST/Stmt.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/Rewrite/Core/Rewriter.h"
+#include <sstream>
+#include "clang/CConv/StructInit.h"
+#include "clang/CConv/MappingVisitor.h"
+
+using namespace llvm;
+using namespace clang;
+
+unsigned int DeclRewriter::getParameterIndex(ParmVarDecl *PV,
+                                             FunctionDecl *FD) {
+  // This is kind of hacky, maybe we should record the index of the
+  // parameter when we find it, instead of re-discovering it here.
+  unsigned int PIdx = 0;
+  for (const auto &I : FD->parameters()) {
+    if (I == PV)
+      return PIdx;
+    PIdx++;
+  }
+  llvm_unreachable("Parameter declaration not found in function declaration.");
+}
+
+void DeclRewriter::rewriteParmVarDecl(const DAndReplace &N) {
+  ParmVarDecl *PV = N.getDecl<ParmVarDecl>();
+
+  // First, find all the declarations of the containing function.
+  DeclContext *DF = PV->getParentFunctionOrMethod();
+  assert(DF != nullptr && "no parent function or method for decl");
+  FunctionDecl *FD = cast<FunctionDecl>(DF);
+
+  // For each function, determine which parameter in the declaration
+  // matches PV, then, get the type location of that parameter
+  // declaration and re-write.
+  unsigned int PIdx = getParameterIndex(PV, FD);
+
+  for (auto *CurFD = FD; CurFD != nullptr; CurFD = CurFD->getPreviousDecl())
+    if (PIdx < CurFD->getNumParams()) {
+      // TODO these declarations could get us into deeper header files.
```

Comment:
So? Won't the source location go to the right file, wherever it is?

Reply:
i honestly don't understand what this comment is getting at. Might be stale?